platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,268 @@
+
+```java
+public interface StateLifecycleManager {
+    /**
+     * Get the mutable state.
+     */
+    State getMutableState();
+
+   /**
+    * Reset the mutable state to the latest immutable state.
+    */
+   void resetMutableState();
+
+   /**
+    * Creates a snapshot for the latest immutable state. 
+    * 
+    * @param targetPath The path to save the snapshot.
+    */
+    State createSnapshot(final @NonNull Path targetPath);
+
+    /**
+     * Creates a mutable copy of the state. The previous mutable state becomes immutable,
+     * replacing the latest immutable state. 
+     * 
+     * @return a mutable copy of the state
+     */
+    State copyMutableState();
+}
+```
+
+#### Notes
+
+- Initial version of the interface included `getLatestImmutableState` method, but upon further consideration, 
+it was removed as it is not necessary. After the state becomes immutable, only three things can happen to it:
+  - It can be hashed
+  - It can be stored to disk as a snapshot
+  - It can become a source for a mutable copy in case of the `resetMutableState` operation
+
+  All these operations are managed by the `StateLifecycleManager` implementation, and therefore the client code does not need to access the immutable state directly.
+- The `resetMutableState` method is to effectively revert all the changes since the last copy was made. This method will require additional code changes
+  as we normally do not allow creation of immutable state copies.
+- The `createSnapshot` method requires the latest immutable state to be hashed. It will block on the `Future` returned by the `State#getHash()` method until the hash is ready.
+
+---
+
+### Implementation Details
+
+The implementation will include the methods defined in the interface and two additional static utility methods:
+
+```java
+public class StateLifecycleManagerImpl implements StateLifecycleManager {
+    /**
+     * Load an immutable copy of a state from the snapshot at the specified path.
+     *
+     * @param sourcePath the path to the snapshot
+     * @return the immutable state loaded from the snapshot
+     */
+    public static State loadSnapshot(Path sourcePath) {
+        // implementation
+    }
+
+    /**
+     * Returns a list of paths to snapshots discovered in the specified directory.
+     *
+     * @param sourcePath the path to the directory
+     * @return a list of paths to snapshots
+     */
+    public static List<Path> discover(Path sourcePath) {
+        // implementation
+    }
+}
+```
+
+
+#### Notes
+
+- The `StateLifecycleManager` implementation will maintain a **1-to-1 relationship** with the state. 
+The assumption is that the application will require only one mutable state at a time, and this state should be initialized during the application's startup.  
+- If client code needs a state loaded from an arbitrary snapshot, it can directly use the `StateLifecycleManagerImpl#loadSnapshot` method, which is `static`. 
```

Comment:
The client code should have any access to the implementation class, only the interface, so I don't see how this will work

Reply:
The client code will need to use the implementation classes anyway. 

However, instead of having static method we can have a constructor that takes a path to the snapshot. And this is how the client code will be able to work with the loaded state.